File: content/snippets/python/interview_lcci/08/01_three_steps.py
def waysToStep(n: int) -> int:
    MOD = 10**9 + 7
    if n == 1: return 1
    if n == 2: return 2
    if n == 3: return 4
    a, b, c = 1, 2, 4
    for _ in range(n - 3):
        a, b, c = b, c, (a + b + c) % MOD
    return c

def main():
    print(waysToStep(3))  # 4
    print(waysToStep(5))  # 13

if __name__ == "__main__":
    main()

File: content/snippets/typescript/interview_lcci/08/01_three_steps.ts
function waysToStep(n: number): number {
    const MOD = 1000000007n;
    if (n === 1) return 1;
    if (n === 2) return 2;
    if (n === 3) return 4;
    let a = 1n, b = 2n, c = 4n;
    for (let i = 3; i < n; i++) {
        [a, b, c] = [b, c, ((a + b) % MOD + c) % MOD];
    }
    return Number(c);
}

console.log(waysToStep(3)); // 4
console.log(waysToStep(5)); // 13
